File: Question/46_Permutations.cpp
class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> ans;
        vector<int> cur;
        vector<bool> used(nums.size(), false);
        dfs(nums, 0, used, cur, ans);
        return ans;
    }
    void dfs(vector<int>& nums, int level, vector<bool>& used, vector<int>& cur, vector<vector<int>>& ans) {
        if(level >= nums.size()) {
            ans.push_back(cur);
            return;
        }

        for(int i = 0; i < nums.size(); ++i) {
            if(used[i] == true) continue;
            cur.push_back(nums[i]);
            used[i] = true;
            dfs(nums, level + 1, used, cur, ans);
            used[i] = false;
            cur.pop_back();
        }
    }
};

File: Question_Java/46_Permutations.java
class Solution {
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> ans = new ArrayList<>();
        dfs(nums, new int[nums.length], new ArrayList<Integer>(), ans);
        return ans;
    }
    private void dfs(int[] nums, int[] used, List<Integer> cur, List<List<Integer>> ans) {
        if (cur.size() == nums.length) {
            ans.add(new ArrayList<Integer>(cur));
            return;
        }
        for (int i = 0; i < nums.length; ++i) {
            if (used[i] == 1) continue;
            used[i] = 1;
            cur.add(nums[i]);
            dfs(nums, used, cur, ans);
            cur.remove(cur.size() - 1);
            used[i] = 0;
        }
    }
}